src/modules/iotjs_module_mqtt.c

```diff
@@ -0,0 +1,719 @@
+    len >>= 7;
+    n++;
+  }
+
+  return n;
+}
+
+
+static uint32_t iotjs_decode_remaining_length(char *buffer, size_t *offset) {
+  unsigned char c;
+  uint32_t remaining_length = 0;
+  uint32_t length = 0;
+  uint32_t multiplier = 1;
+
+  do {
+    if (++length > IOTJS_MODULE_MQTT_MAX_REMAINING_LENGTH_BYTES) {
+      return UINT32_MAX;
+    }
+    c = (unsigned char)buffer[*offset];
+    remaining_length += (c & 127) * multiplier;
+    multiplier *= 128;
+    (*offset)++;
+  } while ((c & 128) != 0);
+
+  return remaining_length;
+}
+
+
+static void iotjs_mqtt_calculate_length(uint16_t *len, uint8_t msb,
+                                        uint8_t lsb) {
+  *len = (msb << 8) | lsb;
+}
+
+
+static uint8_t *iotjs_mqtt_string_serialize(uint8_t *dst_buffer,
+                                            iotjs_bufferwrap_t *src_buffer) {
+  uint16_t len = src_buffer->length;
+  dst_buffer[0] = (uint8_t)(len >> 8);
+  dst_buffer[1] = (uint8_t)(len & 0x00FF);
+  memcpy(dst_buffer + 2, src_buffer->buffer, src_buffer->length);
+  return (dst_buffer + 2 + src_buffer->length);
+}
+
+
+void iotjs_create_ack_callback(char *buffer, char *name, jerry_value_t jsref) {
+  uint8_t packet_identifier_MSB = (uint8_t)buffer[2];
+  uint8_t packet_identifier_LSB = (uint8_t)buffer[3];
+
+  uint16_t package_id = 0;
+  iotjs_mqtt_calculate_length(&package_id, packet_identifier_MSB,
+                              packet_identifier_LSB);
+
+  // The callback takes the packet identifier as parameter.
+  iotjs_jargs_t args = iotjs_jargs_create(1);
+  iotjs_jargs_append_number(&args, package_id);
+
+  jerry_value_t fn = iotjs_jval_get_property(jsref, name);
+  iotjs_make_callback(fn, jsref, &args);
+  jerry_release_value(fn);
+  iotjs_jargs_destroy(&args);
+}
+
+
+JS_FUNCTION(MqttConnect) {
+  DJS_CHECK_THIS();
+
+  DJS_CHECK_ARGS(1, object);
+
+  jerry_value_t joptions = JS_GET_ARG(0, object);
+
+  jerry_value_t jclient_id =
+      iotjs_jval_get_property(joptions, IOTJS_MAGIC_STRING_CLIENTID);
+  jerry_value_t jusername =
+      iotjs_jval_get_property(joptions, IOTJS_MAGIC_STRING_USERNAME);
+  jerry_value_t jpassword =
+      iotjs_jval_get_property(joptions, IOTJS_MAGIC_STRING_PASSWORD);
+  jerry_value_t jkeepalive =
+      iotjs_jval_get_property(joptions, IOTJS_MAGIC_STRING_KEEPALIVE);
+
+  uint8_t connect_flags = 0;
+  uint8_t keep_alive_msb = 0;
+  uint8_t keep_alive_lsb = 10;
+  connect_flags |= MQTT_FLAG_CLEANSESSION;
+
+  uint8_t header_byte = 0;
+  header_byte |= (CONNECT << 4);
+
+  if (!jerry_value_is_undefined(jusername)) {
+    connect_flags |= MQTT_FLAG_USERNAME;
+  }
+  if (!jerry_value_is_undefined(jpassword)) {
+    connect_flags |= MQTT_FLAG_PASSWORD;
+  }
+  if (!jerry_value_is_undefined(jkeepalive)) {
+    uint16_t len = jerry_get_number_value(jkeepalive);
+    keep_alive_msb = (uint8_t)(len >> 8);
+    keep_alive_lsb = (uint8_t)(len & 0x00FF);
+  }
+
+
+  iotjs_bufferwrap_t *client_id = iotjs_bufferwrap_from_jbuffer(jclient_id);
```

Comment:
No type check here. I mentioned that there is a function which creates iotjs string from strings/buffers, or return NULL on error.

Reply:
I don't think it is needed since JS side a buffer is always created for the client id, and that's passed to the C side.